### PR TITLE
[IOTDB-2113]Compaction Refactor and Support aligned timeseries

### PR DESCRIPTION
**Pull request by choubenson**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+      for (Modification modification : compactionUnseqModificationFile.getModifications()) {
+        targetFile.getModFile().write(modification);
+      }
+    }
+    targetFile.getModFile().close();
+  }
+
+  /**
+   * This method is called to recover modifications while an exception occurs during compaction. It
+   * append new modifications of each selected tsfile to its corresponding old mods file and delete
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:362]:
```suggestion
   * appends new modifications of each selected tsfile to its corresponding old mods file and delete
```

File: server/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java [hunk 2]
```diff
@@ -90,6 +90,19 @@ public long assignQueryId(boolean isDataQuery) {
...
   }
 
+  /**
+   * Register a query id for compaction. The name of the compaction thread is
+   * 'pool-x-IoTDB-Compaction-xx', xx in which is usually an integer from 0 to
+   * MAXCOMPACTION_THREAD_NUM. We use the following rules to define query id for compaction: <br>
+   * queryId = xx + Long.MIN_VALUE
+   */
+  public long assignCompactionQueryId() {
+    long threadNum = Long.parseLong((Thread.currentThread().getName().split("-"))[4]);
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/query/control/QueryResourceManager.java:100]: this is unsafe, think about using other way to get an integer

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+      seqTsFileResourceList.remove(resource);
+    }
+    for (TsFileResource resource : selectedUnSeqTsFileResourceList) {
+      unseqTsFileResourceList.remove(resource);
+    }
+  }
+
+  private boolean addReadLock(List<TsFileResource> tsFileResourceList) {
+    for (TsFileResource tsFileResource : tsFileResourceList) {
+      if (tsFileResource.isMerging() | !tsFileResource.isClosed()
```
Comment (qiaojialin) [inline server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:221]:
```suggestion
      if (tsFileResource.isMerging() || !tsFileResource.isClosed()
```

File: .github/workflows/main-unix.yml [hunk 4]
```diff
@@ -9,6 +9,7 @@ on:
       - master
       - 'rel/*'
       - Vector
+      - new-compaction
```
[the remark this answers is not in the document]
Comment (choubenson) [reply .github/workflows/main-unix.yml:12]: resolved

File: .github/workflows/main-unix.yml [hunk 5]
```diff
@@ -17,6 +18,7 @@ on:
       - 'rel/*'
       - cluster_new
       - Vector
+      - new-compaction
```
[the remark this answers is not in the document]
Comment (choubenson) [reply .github/workflows/main-unix.yml:21]: resolved

File: .github/workflows/main-win.yml [hunk 6]
```diff
@@ -15,6 +15,7 @@ on:
       - master
       - 'rel/*'
       - cluster_new
+      - new-compaction
```
[the remark this answers is not in the document]
Comment (choubenson) [reply .github/workflows/main-win.yml:18]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java (hunk 1 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionUtils.java:362]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java (hunk 3 above)
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/task/RewriteCrossSpaceCompactionTask.java:223]: resolved

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java [hunk 7]
```diff
@@ -172,8 +172,11 @@ private IoTDBConstant() {}
   public static final int FILE_NAME_SUFFIX_UNSEQMERGECNT_INDEX = 3;
   public static final String FILE_NAME_SUFFIX_SEPARATOR = "\\.";
 
-  // compaction
-  public static final String COMPACTION_TMP_FILE_SUFFIX = ".target";
+  // inner space compaction
+  public static final String INNER_COMPACTION_TMP_FILE_SUFFIX = ".target";
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java:176]: resolved

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java [hunk 8]
```diff
@@ -172,8 +172,11 @@ private IoTDBConstant() {}
...
+
+  // cross space compaction
+  public static final String CROSS_COMPACTION_TMP_FILE_SUFFIX = ".merge";
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConstant.java:179]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionScheduler.java [hunk 9]
```diff
@@ -136,6 +136,7 @@ private static void doCompactionBalancePriority(
                   virtualStorageGroupName,
                   storageGroupDir,
                   timePartition,
+                  tsFileManager,
                   sequenceFileList,
                   unsequenceFileList,
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionScheduler.java:141]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/AbstractCrossSpaceCompactionTask.java [hunk 10]
```diff
@@ -27,6 +27,8 @@
 
 public abstract class AbstractCrossSpaceCompactionTask extends AbstractCompactionTask {
 
+  public static final String MERGE_SUFFIX = ".merge";
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/AbstractCrossSpaceCompactionTask.java:30]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossCompactionStrategy.java [hunk 11]
```diff
@@ -43,27 +43,25 @@ public AbstractCrossSpaceCompactionTask getCompactionTask(
       String logicalStorageGroupName,
       String virtualStorageGroupName,
       long timePartitionId,
-      CrossSpaceMergeResource mergeResource,
       String storageGroupDir,
+      TsFileManager tsFileManager,
       TsFileResourceList seqTsFileResourceList,
-      TsFileResourceList unSeqTsFileResourceList,
+      TsFileResourceList unseqTsFileResourceList,
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/CrossCompactionStrategy.java:49]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/AbstractCrossSpaceCompactionSelector.java [hunk 12]
```diff
@@ -35,6 +37,7 @@ public AbstractCrossSpaceCompactionSelector(
       String virtualGroupId,
       String storageGroupDir,
       long timePartition,
+      TsFileManager tsFileManager,
       TsFileResourceList sequenceFileList,
       TsFileResourceList unsequenceFileList,
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/AbstractCrossSpaceCompactionSelector.java:42]: resolved

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**